### PR TITLE
Add `linalg.block_diag` and sparse equivalent 

### PR DESCRIPTION
**Pull request by jessegrabowski**
## Description
There are a number of generic math-type operations in `pymc.math` that make more sense in `pytensor`. This PR moves one of them -- `block_diag`, to `pytensor.tensor.slinalg`. There are others that could likely be moved as well.

`block_diag` uses a few support functions that also present opportunities to implement some missing numpy functions in pytensor. I re-wrote `largest_common_dtype` to use `np.promote_types`, but there should probably be a `pt.promote_types` and `pt.result_type`. The latter should just replace this helper function entirely.

There was also a function `ix`, which is an implementation of `np.ix_`. We're missing all of these numpy quick constructors: `np.r_`, `np.c_`, `np.ogrid`, `np.mgrid`. Plus `np.meshgrid`, which is an actual function people use.

`floatX` and `intX` should probably also be pytensor functions instead of PyMC functions.

I tagged this as relevant to #573 because I'm interested in experimenting with linear algebra rewrites out of the `block_diag` `Op`.

## Related Issue

- [ ] Closes #
- [x] Related to #573 

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

**Comment by ricardoV94** [comment]
The floatX functions are actually pretty crappy and should be phased out completely. For instance they fail when you pass a python list of TensorVariables.

PyTensor has tools (including config flags) to handle type promotions that should be relied upon instead or refactored id they don't satisfy out current needs

File: pytensor/tensor/slinalg.py [hunk 1]
```diff
@@ -908,6 +908,115 @@ def solve_discrete_are(A, B, Q, R, enforce_Q_symmetric=False) -> TensorVariable:
...
+        if any(mat.type.ndim != 2 for mat in matrices):
+            raise TypeError("all data arguments must be matrices")
+        if self.sparse:
+            out_type = pytensor.sparse.matrix(self.format, dtype=dtype)
+        else:
+            out_type = pytensor.tensor.matrix(dtype=dtype)
+        return Apply(self, matrices, [out_type])
+
+    def perform(self, node, inputs, output_storage, params=None):
+        dtype = largest_common_dtype(inputs)
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:939]: No need to compute `dtype` in every call, it's available from `node.outputs[0].dtype`

File: pytensor/tensor/slinalg.py [hunk 2]
```diff
@@ -908,6 +908,115 @@ def solve_discrete_are(A, B, Q, R, enforce_Q_symmetric=False) -> TensorVariable:
...
+    ..code-block:: python
+
+        import numpy as np
+        from pytensor.tensor.slinalg import block_diagonal
+
+        matrices = [np.array([[1, 2], [3, 4]]), np.array([[5, 6], [7, 8]])]
+        matrices = [pt.as_tensor_variable(mat) for mat in matrices]
+        result = block_diagonal(matrices)
+
+        print(result)
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:1000]: I imagine this is needed?
```suggestion
        print(result.eval())
```

File: tests/tensor/test_slinalg.py [hunk 3]
```diff
@@ -661,3 +662,14 @@ def test_solve_discrete_are_grad():
...
+
+
+def test_block_diagonal():
+    matrices = [np.array([[1.0, 2.0], [3.0, 4.0]]), np.array([[5.0, 6.0], [7.0, 8.0]])]
+    result = block_diagonal(matrices)
+    np.testing.assert_allclose(result.eval(), scipy.linalg.block_diag(*matrices))
+
+    result = block_diagonal(matrices, format="csr", sparse=True)
+    sp_result = scipy.sparse.block_diag(matrices, format="csr")
+    assert type(result.eval()) == type(sp_result)
```
Comment (ricardoV94) [inline tests/tensor/test_slinalg.py:674]:
```suggestion
    assert isinstance(result.eval()),  type(sp_result))
```

Comment (ricardoV94) [reply tests/tensor/test_slinalg.py:674]: Also store the result instead of calling eval twice?

File: pytensor/tensor/basic.py [hunk 4]
```diff
@@ -4269,6 +4269,25 @@ def take_along_axis(arr, indices, axis=0):
     return arr[_make_along_axis_idx(arr.shape, indices, axis)]
 
 
+def ix_(*args):
```
Comment (ricardoV94) [inline pytensor/tensor/basic.py:4272]: Want to add this to `__all__`?

**Review by ricardoV94 — COMMENTED** [review]
Looks good, just some small tweaks

**Comment by ricardoV94** [comment]
Also are dispatch for JAX/Numba trivial that we could support already?

**Comment by ricardoV94** [comment]
One thing I don't know is mixing the sparse and dense in a single Op. Maybe that's fine.

However, I think the helper function  should follow scipy API, so there would be a `tensor.linalg.block_diag` that builds the dense case (or in this case can just use a pre-built Op) and a `tensor.sparse.block_diag` for the sparse cases.

File: pytensor/tensor/basic.py (hunk 4 above)
Comment (jessegrabowski) [reply pytensor/tensor/basic.py:4272]: I wanted to check with you first

File: pytensor/tensor/slinalg.py (hunk 1 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:939]: Reading `cast_policy`, do we need to do this dtype stuff at all, or can I just delete it and let pytensor figure it out

**Comment by jessegrabowski** [comment]
I think numba/jax should be easy, i'll work on it.

With respect to the API I totally agree about splitting sparse and dense into `pytensor.tensor.slinalg.block_diag` and `pytensor.sparse.block_diag`. I will also refactor the function to match the scipy function, which takes the matrices as `args` instead of a single list.

File: pytensor/tensor/basic.py (hunk 4 above)
Comment (ricardoV94) [reply pytensor/tensor/basic.py:4272]: Doesn't really matter. It shouldn't be too hard right? It would be implemented as classmethod of `ix_`?

**Comment by ricardoV94** [comment]
Can you open an issue in PyMC to deprecate the code there?

File: pytensor/tensor/basic.py (hunk 4 above)
Comment (jessegrabowski) [reply pytensor/tensor/basic.py:4272]: Sorry I got confused, `np.ix_` is a normal function, so we can keep this as-is. If we wanted to do the other ones (`np.r_`, `np.c_`, `np.s_`, `np.ogrid`, `np.mgrid`), they have the funny syntax because they allow fast definition using slices (`np.ogrid[1:2, 1:2]` is the same as `np.ix_([1, 2], [1,2])`.

All of these live in [numpy.lib._index_tricks_impl.py](https://github.com/numpy/numpy/blob/9ab2a257e4f31f3524377d176233d0d4561c8fd4/numpy/lib/_index_tricks_impl.py).

So what I'm trying to say is I have no idea where to put `ix_`.

**Comment by jessegrabowski** [comment]
I'm going to need some handholding with turning it into Blockwise. I think the gfunc_sig should just be `"(n,m)->(o,p)"`, but there are an arbitrary number of inputs all of the same size, so do I need to represent that somehow?

**Comment by ricardoV94** [comment]
Signatures can be defined dynamically when the Op is initialized, that's how we do it for solve.

For this Op you can even use the default signature function we use in the fallback vectorize/blockwise function. I think there's nothing extra to do here?

**Comment by jessegrabowski** [comment]
Yes you're right about sizes

File: pytensor/sparse/basic.py [hunk 5]
```diff
@@ -4260,3 +4256,85 @@ def grad(self, inputs, grads):
 
 
 construct_sparse_from_list = ConstructSparseFromList()
+
+
+class SparseBlockDiagonalMatrix(BaseBlockDiagonal):
+    def make_node(self, *matrices, format: Literal["csc", "csr"] = "csc", name=None):
```
Comment (ricardoV94) [inline pytensor/sparse/basic.py:4262]: The format should be part of init and a prop. It should always be possible to do `out.owner.make_node(*out.owner.inputs)` safely, which wouldn't work here.

This is needed when rebuilding graphs with `clone_replace` for example

File: pytensor/sparse/basic.py [hunk 6]
```diff
@@ -4260,3 +4256,85 @@ def grad(self, inputs, grads):
...
+        matrices = list(map(as_sparse_or_tensor_variable, matrices))
+
+        if any(mat.type.ndim != 2 for mat in matrices):
+            raise TypeError("all data arguments must be matrices")
+
+        out_type = matrix(format=format, dtype=dtype, name=name)
+        return Apply(self, matrices, [out_type])
+
+    def perform(self, node, inputs, output_storage, params=None):
+        format = node.outputs[0].type.format
```
Comment (ricardoV94) [inline pytensor/sparse/basic.py:4275]: After you make format a prop:

```suggestion
        format = self.format
```

File: pytensor/tensor/basic.py (hunk 4 above)
Comment (ricardoV94) [reply pytensor/tensor/basic.py:4272]: Can be here no problem. We might refactor basic another time

File: pytensor/tensor/slinalg.py [hunk 7]
```diff
@@ -908,6 +908,101 @@ def solve_discrete_are(A, B, Q, R, enforce_Q_symmetric=False) -> TensorVariable:
     )
 
 
+def largest_common_dtype(tensors: typing.Sequence[TensorVariable]) -> np.dtype:
+    return ft.reduce(lambda l, r: np.promote_types(l, r), [x.dtype for x in tensors])
+
+
+def block_diag_grad(inputs, gout):
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:915]: Why did you refactor?

File: pytensor/tensor/slinalg.py [hunk 8]
```diff
@@ -908,6 +908,101 @@ def solve_discrete_are(A, B, Q, R, enforce_Q_symmetric=False) -> TensorVariable:
...
+    def grad(self, inputs, gout):
+        return block_diag_grad(inputs, gout)
+
+    def infer_shape(self, fgraph, nodes, shapes):
+        first, second = zip(*shapes)
+        return [(pt.add(*first), pt.add(*second))]
+
+
+class BlockDiagonalMatrix(BaseBlockDiagonal):
+    def make_node(self, *matrices, name=None):
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:939]: Never saw name in make_node, can it be reached from `op.__call__`? That would actually be a good place to allow a name kwarg, so it would work with all Ops!

File: tests/link/numba/test_slinalg.py [hunk 9]
```diff
@@ -102,3 +102,30 @@ def test_solve_triangular_raises_on_nan_inf(value):
         ValueError, match=re.escape("Non-numeric values (nan or inf) returned ")
     ):
         f(A_tri, b)
+
+
+def test_block_diag():
```
Comment (ricardoV94) [inline tests/link/numba/test_slinalg.py:107]: Isn't there a compare py_and_numba?

**Comment by ricardoV94** [comment]
Jax will work with vmap only need to dispatch the base case.

Numba doesn't yet have support for Blockwise. We should be able to do something simple woth guvectorize although it has some limitations like only working with single outputs

File: pytensor/tensor/slinalg.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/tensor/slinalg.py]: I don't know what's going on here, but we need to specify the dtype of the output in make_node. I think the helper was used to figure that out

File: pytensor/tensor/slinalg.py (hunk 7 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:915]: This was an intermediate step I didn't delete, i'll take it out.

File: pytensor/tensor/slinalg.py (hunk 8 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:939]: It breaks `Blockwise` but works otherwise

File: pytensor/sparse/basic.py (hunk 5 above)
Comment (jessegrabowski) [reply pytensor/sparse/basic.py:4262]: Ok -- I thought I was being clever by delaying the need to commit to a format as late as possible.

File: pytensor/tensor/slinalg.py (hunk 8 above)
Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:939]: What breaks Blockwise?

Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:939]: Having kwargs in `make_node`

Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:939]: make_node shouldn't have kwargs I think, but Op.__call__ can and name would be a neat one to add?

Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:939]: Oh interesting! I'll have a look at that. I really like having names for the stack traces and dprints, so i'd like it if every Op had it.

Comment (ricardoV94) [reply pytensor/tensor/slinalg.py]: Me too, it's also awkward to need two lines to give a name

**Comment by jessegrabowski** [comment]
I guess we can't blockwise Sparse Ops? Blockwise uses `pt.as_tensor`, which breaks if the inputs are sparse. Will need to add some logic to handle that case. This would be the first function it would work with, I think. Otherwise, though, I think this is done.

File: pytensor/tensor/slinalg.py [hunk 10]
```diff
@@ -908,6 +908,102 @@ def solve_discrete_are(A, B, Q, R, enforce_Q_symmetric=False) -> TensorVariable:
...
 
 
+def largest_common_dtype(tensors: typing.Sequence[TensorVariable]) -> np.dtype:
+    return ft.reduce(lambda l, r: np.promote_types(l, r), [x.dtype for x in tensors])
+
+
+class BaseBlockDiagonal(Op):
+    __props__ = ("gufunc_signature",)
+
+    def __init__(self, n_inputs):
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:918]: Could save `self.n_inputs` and use in `make_node` where we now check `if not matrices` and instead check that we get the promised number of inputs?

You can do the assert > 0 here already

File: pytensor/tensor/slinalg.py [hunk 11]
```diff
@@ -908,6 +908,102 @@ def solve_discrete_are(A, B, Q, R, enforce_Q_symmetric=False) -> TensorVariable:
...
+
+        out_type = pytensor.tensor.matrix(dtype=dtype)
+        return Apply(self, matrices, [out_type])
+
+    def perform(self, node, inputs, output_storage, params=None):
+        dtype = node.outputs[0].type.dtype
+        output_storage[0][0] = scipy.linalg.block_diag(*inputs).astype(dtype)
+
+
+def block_diag(*matrices: TensorVariable, name=None):
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:958]:
```suggestion
def block_diag(*matrices: TensorVariable):
```

File: pytensor/tensor/slinalg.py [hunk 12]
```diff
@@ -1,3 +1,4 @@
+import functools as ft
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:1]: Nitpick, is this a common alias?

```suggestion
from functools import reduce
```

File: pytensor/tensor/slinalg.py [hunk 13]
```diff
@@ -908,6 +908,102 @@ def solve_discrete_are(A, B, Q, R, enforce_Q_symmetric=False) -> TensorVariable:
...
+    Given the inputs `A`, `B` and `C`, the output will have these arrays arranged on the diagonal:
+
+    [[A, 0, 0],
+     [0, B, 0],
+     [0, 0, C]]
+
+    Parameters
+    ----------
+    A, B, C ... : tensors
+        Input matrices to form the block diagonal matrix. Each matrix should have the same number of dimensions, and the
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:971]: This is not correct. Blockwise accepts different number of batch dims and also broadcasts when they have length 1.

File: pytensor/tensor/slinalg.py [hunk 14]
```diff
@@ -908,6 +908,102 @@ def solve_discrete_are(A, B, Q, R, enforce_Q_symmetric=False) -> TensorVariable:
...
+            for i in range(len(inputs))
+        ]
+        return [gout[0][slc] for slc in slices]
+
+    def infer_shape(self, fgraph, nodes, shapes):
+        first, second = zip(*shapes)
+        return [(pt.add(*first), pt.add(*second))]
+
+
+class BlockDiagonalMatrix(BaseBlockDiagonal):
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:940]:
```suggestion
class BlockDiagonal(BaseBlockDiagonal):
```

Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:940]: Any reason for the Matrix suffix or was it there originally?

File: pytensor/tensor/slinalg.py [hunk 15]
```diff
@@ -908,6 +908,102 @@ def solve_discrete_are(A, B, Q, R, enforce_Q_symmetric=False) -> TensorVariable:
...
+        The block diagonal matrix formed from the input matrices.
+
+    Examples
+    --------
+    Create a block diagonal matrix from two 2x2 matrices:
+
+    ..code-block:: python
+
+        import numpy as np
+        from pytensor.tensor.slinalg import block_diag
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:988]: The slinalg nlinalg all are available and linalg which is the user facing one

```suggestion
        from pytensor.tensor.linalg import block_diag
```

File: pytensor/tensor/slinalg.py [hunk 16]
```diff
@@ -908,6 +908,102 @@ def solve_discrete_are(A, B, Q, R, enforce_Q_symmetric=False) -> TensorVariable:
     )
 
 
+def largest_common_dtype(tensors: typing.Sequence[TensorVariable]) -> np.dtype:
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:911]: This in an odd place for this function? Maybe make it a classmethod of the Op for the time being?

Wouldn't want people finding and importing it from here. Can also put it in a better place instead but that's more work

**Review by ricardoV94 — COMMENTED** [review]
Only nitpicks left, address what you want and close what you don't!

File: pytensor/sparse/basic.py [hunk 17]
```diff
@@ -4260,3 +4256,90 @@ def grad(self, inputs, grads):
 
 
 construct_sparse_from_list = ConstructSparseFromList()
+
+
+class SparseBlockDiagonalMatrix(BaseBlockDiagonal):
```
Comment (ricardoV94) [inline pytensor/sparse/basic.py:4261]: Same question here
```suggestion
class SparseBlockDiagonal(BaseBlockDiagonal):
```

**Comment by ricardoV94** [comment]
> I guess we can't blockwise Sparse Ops? Blockwise uses `pt.as_tensor`, which breaks if the inputs are sparse. Will need to add some logic to handle that case. This would be the first function it would work with, I think. Otherwise, though, I think this is done.

We don't have a sparse type with `ndim != 2` so that's only the tip of the iceberg.

File: pytensor/tensor/slinalg.py (hunk 16 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:911]: I was waiting for a resolution on the other convo thread about whether it's even needed

File: pytensor/sparse/basic.py [hunk 18]
```diff
@@ -4260,3 +4256,90 @@ def grad(self, inputs, grads):
...
+
+    def perform(self, node, inputs, output_storage, params=None):
+        dtype = node.outputs[0].type.dtype
+        output_storage[0][0] = scipy.sparse.block_diag(
+            inputs, format=self.format
+        ).astype(dtype)
+
+
+def block_diag(
+    *matrices: TensorVariable, format: Literal["csc", "csr"] = "csc", name=None
```
[the remark this answers is not in the document]
Comment (jessegrabowski) [reply pytensor/sparse/basic.py:4288]: No just sloppy : )

File: pytensor/tensor/slinalg.py (hunk 12 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:1]: Not sure, I think I saw it in the pytensor code once. But you're right that for base library it's cleaner to just import the functions directly.

File: pytensor/tensor/slinalg.py (hunk 13 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:971]: I got errors when I tried matrices with different batch dims, but I didn't try broadcasting to dimensions with size 1.

File: pytensor/tensor/slinalg.py (hunk 14 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:940]: It was there originally, but I agree it's too wordy

File: pytensor/tensor/slinalg.py (hunk 16 above)
Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:911]: We still need to define a dtype for the output.

File: pytensor/tensor/slinalg.py (hunk 13 above)
Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:971]: Do you still see errors? Blockwise should introduce expand dims, so the only failure case would be broadcasting?

File: pytensor/tensor/slinalg.py (hunk 16 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py]: I renamed it to `_largest_common_dtype` and it's not in `__all__` so if a user finds it kudos to him I guess

File: pytensor/tensor/slinalg.py (hunk 13 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py]: Broadcasting works, I added a test for it. It was failing when I tried different batch sizes, which doesn't make sense anyway I think.

File: pytensor/sparse/basic.py [hunk 19]
```diff
@@ -4260,3 +4256,82 @@ def grad(self, inputs, grads):
 
 
 construct_sparse_from_list = ConstructSparseFromList()
+
+
+class SparseBlockDiagonal(BaseBlockDiagonal):
+    __props__ = ("format",)
```
Comment (ricardoV94) [inline pytensor/sparse/basic.py:4262]: Big gotcha you need to add the props of the base class as well

File: pytensor/tensor/slinalg.py [hunk 20]
```diff
@@ -908,6 +908,105 @@ def solve_discrete_are(A, B, Q, R, enforce_Q_symmetric=False) -> TensorVariable:
     )
 
 
+def _largest_common_dtype(tensors: typing.Sequence[TensorVariable]) -> np.dtype:
+    return reduce(lambda l, r: np.promote_types(l, r), [x.dtype for x in tensors])
+
+
+class BaseBlockDiagonal(Op):
+    __props__ = ("gufunc_signature", "n_inputs")
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:916]: Not all variables need to be props. Usually we make props the ones needed to initialize the Op (for instance the inplace rewrite).

The real contract for props, however, is that if two Ops have the same type, same inputs and same props, then they can be considered equivalent and merged. For this use case we actually don't need any props.

Still for the use case of more easily reconstructing an Op I would include n_inputs only. This is also used for the automatic `__repr__`

Another reason not to include the signature is that it's always a prop of the Blockwise and would show twice in the dprint when you have a Blockwise BlockDiagonal 

```suggestion
    __props__ = ("n_inputs")
```

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
Looks pretty neat! Two not so nitpick comments. 

Also nice catch with the missing test for grad

File: pytensor/sparse/basic.py (hunk 19 above)
Comment (jessegrabowski) [reply pytensor/sparse/basic.py:4262]: I thought they get set union'd when you subclass?

File: pytensor/tensor/slinalg.py (hunk 13 above)
Comment (ricardoV94) [reply pytensor/tensor/slinalg.py]: What do you mean different batch sizes? Blockwise adds expand dims automatically to align the number of batch dims, so that shouldn't be possible?

File: pytensor/sparse/basic.py (hunk 19 above)
Comment (ricardoV94) [reply pytensor/sparse/basic.py:4262]: Maybe! Would be nice if we could do that but I don't know how.

We could also investigate using slots instead

File: pytensor/tensor/slinalg.py (hunk 13 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py]: This errors:

```
    # Different batch sizes
    A = np.random.normal(size=(batch_size + 3, 2, 2)).astype(config.floatX)
    B = np.random.normal(size=(batch_size, 4, 4)).astype(config.floatX)
    result = block_diag(A, B).eval()
```

with:

```
E           ValueError: Incompatible Blockwise batch input shapes [(8, 2, 2), (5, 4, 4)]
```

**Review by ricardoV94 — APPROVED** [review]

File: pytensor/tensor/slinalg.py (hunk 13 above)
Comment (ricardoV94) [reply pytensor/tensor/slinalg.py]: That's invalid, batch shapes must be broadcastable. 8 and 5 are not broadcastable
